core/src/main/java/org/elasticsearch/cluster/MasterNodeChangePredicate.java

```diff
@@ -33,13 +33,13 @@ private MasterNodeChangePredicate() {
      */
     public static Predicate<ClusterState> build(ClusterState currentState) {
         final long currentVersion = currentState.version();
-        final String currentMaster = currentState.nodes().getMasterNodeId();
+        final String currentMasterId = currentState.nodes().getMasterNode().getEphemeralId();
         return newState -> {
-            final String newMaster = newState.nodes().getMasterNodeId();
+            final String newMasterId = newState.nodes().getMasterNode().getEphemeralId();
             final boolean accept;
-            if (newMaster == null) {
+            if (newMasterId == null) {
                 accept = false;
-            } else if (newMaster.equals(currentMaster) == false){
+            } else if (newMasterId.equals(currentMasterId) == false){
```

Comment:
That's I had it before. I refactored it as I prefer to have explicit references to places that depends on the ephemeral id. I know we use the disco nodes themselves in many places but I don't want to add another one.